lambda-buffers-proto/compiler.proto

```diff
@@ -491,3 +489,107 @@ message CompilerInput {
   // Modules to compile.
   repeated Module modules = 1;
 }
+
+
+// Compiler Failure can be extended with other classes of errors.
+message CompilerFailure {
+  // Reason why the compilation failed.
+  oneof failure {
+    KindCheckErr kc_err = 1;
+  }
+
+}
+
+/* Inference Failure 
+ */
+
+message InferenceErr {
+
+  // Error referring to an unbound term. This usually means that the term was
+  // not defined.
+  message UnboundTermErr {
+    string msg = 1; 
+  }
+
+  // This should never happen.
+  message ImpossibleErr {
+    string msg = 1;
+  }
+
+  // Unification has failed - type is incorrectly defined.
+  message UnificationErr {
+    string msg = 1; 
+  }
+
+  // Infinite recursion due to a substitution replacing a term with a term
+  // containing itself. Should never type check.
+  message RecursiveSubstitutionErr {
+    string msg = 1;
+  }
+
+  // The types of inference errors.
+  oneof inf_err {
+    UnboundTermErr unbound_term_e = 1;
+    ImpossibleErr impossible_e = 2;
+    UnificationErr unification_e = 3;
+    RecursiveSubstitutionErr recursive_subs_e = 4; 
+  }
+}
+
+// The inferred type differs from the type as defined.
+message InconsistentTypeErr {
+  TyDef type_def = 1;
+}
+
+// Inference failed.
+message InferenceFailure {
+  TyDef type_def = 1;
+  InferenceErr infer_err = 2;
+}
+
+
+/* Kindcheck Failure
+
+   Information regarding a Kindcheck failure.
+ */
+message KindCheckErr {
+
+  oneof kind_check_error {
+    InferenceFailure inference_failure_err = 1;
+    InconsistentTypeErr inconsistent_type_err = 2; 
+  }
+}
+
+// A namespaced type term.
+message NSTypeName {
+  string namespaced_name = 1;
+}
+
+// A validated type definition.
+message TypeDefinition {
+  NSTypeName ns_name = 1 ;
+  Kind kind = 2;
+}
+
+/*
+  Compiler Output
+
+  A successful Compilation Output.
+ */
+message CompilerOutput {
+  // A map of checked type definitions.
+  repeated TypeDefinition type_defs = 1;
```

Comment:
The problem with the `TyDef` as it stands is that it holds too little information for it to be relevant (i.e. the module info is one level above - which is what would be relevant). Maybe there is a way to include it in the `Module`. Otherwise I thought this would be the nicest way to get the info.